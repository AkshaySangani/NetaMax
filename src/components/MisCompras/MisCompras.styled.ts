import styled from "styled-components";
import { responsive } from "styled/responsive";

import { Box } from "@chakra-ui/react";

export const MisComprasContainer = styled(Box)`
  background: white;
  border: 0.5px solid #9f9f9f;
  box-sizing: border-box;
  box-shadow: 0px 2px 6px rgba(142, 142, 142, 0.25);
  border-radius: 14px;
  display: flex;
  height: 414px;
  flex-direction: column;

  ${responsive.mdPlus} {
    flex-direction: column;
    height: 188px;
  }
`;

export const ButtonsContainer = styled(Box)`
  padding: 10px 20px;
  display: flex;
  justify-content: space-evenly !important;

  ${responsive.mdPlus} {
    display: none;
  }
`;
