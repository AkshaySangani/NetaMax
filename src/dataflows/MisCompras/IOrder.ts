import { IProduct } from "dataflows/Product/IProduct";

export interface IOrder {
  /**
   * The order id
   * @type {string}
   */
  id: string;

  /**
   * The order status id
   * @type {string}
   */
  orderStatusId: string;

  /**
   * The sipping status id
   * @type {string}
   */
  shippingStatusId: string;

  /**
   * the current estimated time of arrival
   * @type {string}
   **/
  currentETA: string;

  /**
   * The total amount of the order
   * @type {string}
   **/
  //TODO: Backend should pass this as number.
  orderTotal: string;

  /**
   * The store name where the order was placed
   * @type {string}
   **/
  storeName: string;

  /**
   * The pick up token
   * @type {number}
   */
  pickUpToken: number;

  /**
   * The products included in the order
   * @type {{total: number, products: Pick<IProduct, "id" | "name" | "seoFilename">[]}}
   **/
  productOrders: {
    /**
     * The total number of products in the order
     * @type {number}
     */
    //TODO: Why we do need this? is it not the same that products.length?
    total: number;

    /**
     * The products included in the order
     * @type {Pick<IProduct, "id" | "name" | "seoFilename">[]}
     **/
    products: Pick<IProduct, "id" | "name" | "seoFilename">[];
  };
}
