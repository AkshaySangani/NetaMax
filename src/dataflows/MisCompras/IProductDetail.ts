import { IProduct } from "dataflows/Product/IProduct";

export interface IProductDetail extends IProduct {
  /**
   * The amout of products
   * @type {string}
   */
  quantity: string;
}
