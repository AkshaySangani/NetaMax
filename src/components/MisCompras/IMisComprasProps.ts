import { IOrder } from "dataflows/MisCompras/IOrder";

export interface IMisComprasProps {
  /**
   * The customer's orders.
   * @type {IOrder[]}
   */
  orders?: IOrder[];

  /**
   * Indicates if there are more orders to load.
   * @type {boolean}
   */
  hasMoreOrders: boolean;

  /**
   * load the next page of orders
   * @type {() => void}
   **/
  loadNextPage: () => void;
}
