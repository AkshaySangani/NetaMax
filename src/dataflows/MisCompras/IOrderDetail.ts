import { IProductDetail } from "./IProductDetail";

export interface IOrderDetail {
  /**
   * The total amount of the order.
   * @type {number}
   */
  orderTotal: number;

  /**
   * The delivery date.
   * @type {string}
   */
  currentETA: string;

  /**
   * The company name.
   * @type {string}
   */
  name: string;

  /**
   * The subtotal without discounts.
   * @type {number}
   */
  orderSubtotalInclTax: number;

  /**
   * The order discount
   * @type {number}
   */
  orderDiscount: number;

  /**
   * The company address
   * @type {string}
   */
  companyAddress: string;

  /**
   * The order id
   * @type {string}
   */
  id: string;

  /**
   * The store phone number
   * @type {string}
   */
  companyPhoneNumber: string;

  /**
   * The products order.
   * @type {IProduct[]}
   */
  productsOrder: IProductDetail[];

  /**
   * The order status id
   * @type {string}
   */
  orderStatusId: string;
}
