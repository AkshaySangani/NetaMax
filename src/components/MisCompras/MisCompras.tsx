import { ReactElement } from "react";

import InfiniteScroll from "react-infinite-scroll-component";

import { EmbeddedVideo } from "components/common/EmbeddedVideoComponent/EmbeddedVideo";
import { Loader } from "components/common/Loader";
import { OrderCard } from "components/common/OrderCard";
import {
  TrackerApp,
  TAB_MISCOMPRAS_VIDEO_PlAYED,
  MIS_COMPRAS_NO_PURCHASES_BUTTON_CLICKED,
} from "constants/amplitudeConstants";
import router from "next/router";
import { trackEvent } from "utils/tracker";

import { Box, Button, Center, Heading, Text } from "@chakra-ui/react";

import { IMisComprasProps } from "./IMisComprasProps";

/**
 * The MisCompras component.
 * @param {IMisComprasProps} props the component props.
 * @returns {ReactElement} the component element.
 */
export const MisCompras = (props: IMisComprasProps): ReactElement => {
  const { orders, hasMoreOrders, loadNextPage } = props;
  const hasOrders = Array.isArray(orders) && orders.length > 0;

  return hasOrders ? (
    <>
      <InfiniteScroll
        dataLength={orders.length}
        style={{ overflow: "hidden" }}
        next={loadNextPage}
        hasMore={hasMoreOrders}
        loader={
          <Center pt="30px">
            <Loader />
          </Center>
        }
      >
        {orders.map((order) => {
          return (
            <OrderCard
              key={order.id}
              orderId={order.id}
              storeName={order.storeName}
              date={new Date(order.currentETA)}
              totalAmount={parseFloat(order.orderTotal)}
              productImages={order.productOrders.products.map((product) => product.seoFilename)}
              orderStatusId={order.orderStatusId}
              totalProducts={order.productOrders.total}
              pickUpToken={order.pickUpToken}
            />
          );
        })}
      </InfiniteScroll>
    </>
  ) : (
    <Box pt="30px">
      <Center height="100%" width="100%" padding="8" paddingTop="6" gap="10" flexDirection="column">
        <EmbeddedVideo
          videoUrlId="D-OJLyECbfo"
          onPlayVideoClick={() => {
            trackEvent(TAB_MISCOMPRAS_VIDEO_PlAYED, [TrackerApp.Amplitude, TrackerApp.Segment]);
          }}
        />

        <Heading data-testid="tutorial-message" fontWeight="800" fontSize="20px">
          No tienes compras a√∫n :(
        </Heading>

        <Box mt="-20px" textAlign="center" width="80%">
          <Text lineHeight="20px" fontSize="20px" color="#4A4A4A">
            Te invitamos a conocer nuestros productos mas baratos que cualquier super
          </Text>
        </Box>
        <Button
          width="290px"
          height="56px"
          borderRadius="20px"
          colorScheme="netaBlue"
          fontSize="18px"
          onClick={() => {
            trackEvent(MIS_COMPRAS_NO_PURCHASES_BUTTON_CLICKED, [
              TrackerApp.Amplitude,
              TrackerApp.Segment,
            ]);

            router.push("/");
          }}
          data-testid="redirect-button"
        >
          Ir a comprar
        </Button>
      </Center>
    </Box>
  );
};
