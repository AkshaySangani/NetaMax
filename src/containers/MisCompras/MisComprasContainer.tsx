import React, { useEffect, ReactElement } from "react";

import { useDispatch, useSelector } from "react-redux";

import { BackToHome } from "components/common/BackHomeComponent/BackToHome";
import { GoToLogin } from "components/common/GoToLogin/GoToLogin";
import { Loader } from "components/common/Loader";
import { IMisComprasProps } from "components/MisCompras/IMisComprasProps";
import { MisCompras } from "components/MisCompras/MisCompras";
import { HeaderContainer } from "components/ProductDetails/ProductDetails.styled";
import {
  MI_CUENTA_MIS_COMPRAS_PAGE_VIEWED,
  MIS_COMPRAS_NO_PURCHASES_PAGE_VIEWED,
  TrackerApp,
} from "constants/amplitudeConstants";
import { selectCustomer, selectIsLoggedIn } from "dataflows/Auth/AuthSelectors";
import { onOpen as onOpenAction } from "dataflows/Checkout/CheckoutSlice";
import {
  selectIsAmplitudeInitialized,
  selectIsSegmentInitialized,
  selectIsSegmentNodeInitialized,
} from "dataflows/EventTracking/EventTrackingSelectors";
import {
  selectIsLoadingOrders,
  selectOrders,
  selectOrdersMetadata,
} from "dataflows/MisCompras/MisComprasSelectors";
import { getOrders } from "dataflows/MisCompras/MisComprasThunks";
import { useIsMounted } from "hooks/useIsMounted";
import { trackEvent } from "utils/tracker";

import { Center, Container } from "@chakra-ui/react";

/**
 * The MisComprasContainer component.
 * @returns {ReactElement} the component element.
 */
export const MisComprasContainer = (): ReactElement => {
  const isLoggedIn = useSelector(selectIsLoggedIn);
  const dispatch = useDispatch();
  const orders = useSelector(selectOrders);
  const metadata = useSelector(selectOrdersMetadata);
  const isLoadingOrders = useSelector(selectIsLoadingOrders);
  const userAuth = useSelector(selectCustomer);
  const isAmplitudeInitialized = useSelector(selectIsAmplitudeInitialized);
  const isSegmentInitialized = useSelector(selectIsSegmentInitialized);
  const isSegmentNodeInitialized = useSelector(selectIsSegmentNodeInitialized);
  const isMounted = useIsMounted();

  /**
   * handles OnOpen action login.
   * @returns {void}
   */
  const onOpenLoginFlow = () => {
    dispatch(onOpenAction({ onlyLoginSteps: true }));
  };

  useEffect(() => {
    if (isAmplitudeInitialized && isSegmentInitialized && isSegmentNodeInitialized) {
      trackEvent(MI_CUENTA_MIS_COMPRAS_PAGE_VIEWED, [TrackerApp.Amplitude, TrackerApp.Segment]);
    }
  }, [isAmplitudeInitialized, isSegmentInitialized, isSegmentNodeInitialized]);

  useEffect(() => {
    if (userAuth) {
      dispatch(
        getOrders({
          npp: metadata.perPage,
          page: metadata.currentPage,
          userId: userAuth.id,
        })
      );
    }
  }, [isLoggedIn]);

  useEffect(() => {
    if (
      isMounted &&
      isLoggedIn &&
      !isLoadingOrders &&
      isAmplitudeInitialized &&
      isSegmentInitialized &&
      isSegmentNodeInitialized &&
      orders &&
      orders?.length == 0
    ) {
      trackEvent(MIS_COMPRAS_NO_PURCHASES_PAGE_VIEWED, [TrackerApp.Amplitude, TrackerApp.Segment]);
    }
  }, [
    orders,
    isLoadingOrders,
    isAmplitudeInitialized,
    isSegmentInitialized,
    isSegmentNodeInitialized,
  ]);

  const hasMoreOrders = metadata.totalPages > metadata.currentPage;

  /**
   * Action to fetch more orders.
   * @returns {void}
   */
  const loadNextPage = () => {
    if (userAuth) {
      dispatch(
        getOrders({
          npp: metadata.perPage,
          page: metadata.currentPage + 1,
          userId: userAuth.id,
        })
      );
    }
  };

  const misComprasProps: IMisComprasProps = {
    orders,
    hasMoreOrders,
    loadNextPage,
  };

  if (!isLoggedIn) return <GoToLogin onOpenLogin={onOpenLoginFlow} />;

  return isLoadingOrders && orders?.length === 0 ? (
    <Center h="50%" w="100%" position="fixed">
      <Loader />
    </Center>
  ) : (
    <Container maxW="container.xl">
      <HeaderContainer>
        <BackToHome size={18} weight={700} />
      </HeaderContainer>

      <MisCompras {...misComprasProps} />
    </Container>
  );
};
