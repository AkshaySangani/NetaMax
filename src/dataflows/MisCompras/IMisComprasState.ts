import { IPagedData } from "dataflows/IPagedData";

import { IOrder } from "./IOrder";
import { IOrderDetail } from "./IOrderDetail";

export interface IMisComprasState {
  /**
   * Flag that indicates if the order is loading.
   * @type {boolean}
   */
  isLoadingOrders: boolean;

  /**
   * User's orders.
   * @type {IOrder[]}
   */
  orders: IPagedData<IOrder>;

  /**
   * Flag that indicates if the order is loading.
   * @type {boolean}
   */
  isLoadingGetOrderByOrderId: boolean;

  /**
   * User's order.
   * @type {IOrderDetail}
   */
  order?: IOrderDetail;
}
