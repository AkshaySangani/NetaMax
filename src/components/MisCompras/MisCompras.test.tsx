import { cleanup, render, screen } from "@testing-library/react";
import { MisCompras } from "./MisCompras";

const loadNextPage = jest.fn();

describe("Render MisCompras component", () => {
  it("MisCompras should be render correctly with no orders and a tutorial video", () => {
    const { container } = render(
      <MisCompras hasMoreOrders={false} loadNextPage={loadNextPage} orders={[]} />
    );
    expect(container).toMatchSnapshot();
    expect(screen.getByTestId("tutorial-message").textContent).toMatch("No tienes compras aÃºn :(");
    expect(screen.getByTestId("redirect-button")).toBeTruthy();
    expect(screen.getByTestId("redirect-button").textContent).toMatch("Ir a comprar");
  });

  it("MisCompras should be render correctly with orders", () => {
    const { container } = render(
      <MisCompras
        hasMoreOrders={false}
        loadNextPage={loadNextPage}
        orders={[
          {
            id: "1",
            orderStatusId: "1",
            shippingStatusId: "1",
            currentETA: "2022-03-23T18:00:00.000Z",
            orderTotal: "20",
            storeName: "Abarrotes",
            pickUpToken: 123123,
            productOrders: {
              total: 1,
              products: [
                {
                  id: "1",
                  name: "Leche",
                  seoFilename: "leche.jpg",
                },
              ],
            },
          },
          {
            id: "2",
            orderStatusId: "2",
            shippingStatusId: "2",
            currentETA: "2022-03-24T18:00:00.000Z",
            orderTotal: "20",
            storeName: "Abarrotes Tienda leo",
            pickUpToken: 123123,
            productOrders: {
              total: 1,
              products: [
                {
                  id: "2",
                  name: "Coca",
                  seoFilename: "leche.jpg",
                },
              ],
            },
          },
        ]}
      />
    );
    expect(container).toMatchSnapshot();
  });

  afterEach(() => {
    cleanup();
  });
});
