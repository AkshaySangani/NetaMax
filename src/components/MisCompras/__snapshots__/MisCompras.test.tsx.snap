// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render MisCompras component MisCompras should be render correctly with no orders and a tutorial video 1`] = `
<div>
  <div
    class="css-npgrt2"
  >
    <div
      class="css-dd7c02"
    >
      <div
        class=""
        style="height: 100%; top: 100%;"
      >
        <div
          class=""
          id=""
        />
      </div>
      <h2
        class="chakra-heading css-1rjsx0y"
        data-testid="tutorial-message"
      >
        No tienes compras aún :(
      </h2>
      <div
        class="css-huoxe1"
      >
        <p
          class="chakra-text css-1dp9ibz"
        >
          Te invitamos a conocer nuestros productos mas baratos que cualquier super
        </p>
      </div>
      <button
        class="chakra-button css-19i0htf"
        data-testid="redirect-button"
        type="button"
      >
        Ir a comprar
      </button>
    </div>
  </div>
</div>
`;

exports[`Render MisCompras component MisCompras should be render correctly with orders 1`] = `
<div>
  <div
    class="infinite-scroll-component__outerdiv"
  >
    <div
      class="infinite-scroll-component "
      style="height: auto; overflow: hidden;"
    >
      <div
        class="sc-bdvvtL iRDqbn css-0"
      >
        <div
          class="css-9y56jl"
        >
          <div
            class="css-f1o79g"
          >
            <div
              class="css-k008qs"
            >
              <p
                class="chakra-text css-13ih39a"
              >
                Tiendita:
              </p>
              <p
                class="chakra-text css-1wbddnu"
              >
                Abarrotes
              </p>
            </div>
            <div
              class="css-k008qs"
            >
              <p
                class="chakra-text css-13ih39a"
              >
                Fecha de entrega:
              </p>
              <p
                class="chakra-text css-1wbddnu"
              >
                Miércoles, Marzo 23, 2022
              </p>
            </div>
            <div
              class="css-k008qs"
            >
              <p
                class="chakra-text css-13ih39a"
              >
                Total:
              </p>
              <p
                class="chakra-text css-1wbddnu"
              >
                $20.00
              </p>
            </div>
            <div
              class="css-k008qs"
            >
              <p
                class="chakra-text css-13ih39a"
              >
                Código de entrega:
              </p>
              <p
                class="chakra-text css-1wbddnu"
              >
                123123
              </p>
            </div>
          </div>
          <hr
            aria-orientation="horizontal"
            class="chakra-divider css-wz74kr"
          />
          <div
            class="css-1hw8qak"
          >
            <div
              class="css-mz9qt7"
            >
              <div
                class="css-131wcsq"
              >
                <p
                  class="chakra-text css-d8f8eb"
                  data-testid="order-header"
                >
                  Estado del pedido
                </p>
              </div>
            </div>
          </div>
          <hr
            aria-orientation="horizontal"
            class="chakra-divider css-wz74kr"
          />
          <div
            class="css-f1o79g"
          >
            <div
              class="css-13ih39a"
            >
              Tus productos:
            </div>
            <div
              class="css-15qbb6"
            >
              <div
                class="css-18vhmwz"
              >
                <img
                  class="chakra-image__placeholder css-1jtd0sc"
                  src="/assets/images/noimg.png"
                />
              </div>
            </div>
            <hr
              aria-orientation="horizontal"
              class="chakra-divider css-1p98don"
            />
          </div>
          <div
            class="sc-gsDKAQ eMaAuu css-0"
          >
            <button
              class="chakra-button css-6d4n7v"
              type="button"
            >
              Ver detalle del pedido
            </button>
            <hr
              aria-orientation="vertical"
              class="chakra-divider css-12gj880"
            />
            <a
              class="chakra-link css-0"
              href="https://api.whatsapp.com/send/?phone=%2B525545439866&text=Hola%2CTengo+una+duda..&app_absent=0"
              target="_blank"
            >
              <button
                class="chakra-button css-6d4n7v"
                type="button"
              >
                Ayuda
              </button>
            </a>
          </div>
        </div>
        <hr
          aria-orientation="horizontal"
          class="chakra-divider css-18kxs9k"
        />
      </div>
      <div
        class="sc-bdvvtL iRDqbn css-0"
      >
        <div
          class="css-9y56jl"
        >
          <div
            class="css-f1o79g"
          >
            <div
              class="css-k008qs"
            >
              <p
                class="chakra-text css-13ih39a"
              >
                Tiendita:
              </p>
              <p
                class="chakra-text css-1wbddnu"
              >
                Abarrotes Tienda leo
              </p>
            </div>
            <div
              class="css-k008qs"
            >
              <p
                class="chakra-text css-13ih39a"
              >
                Fecha de entrega:
              </p>
              <p
                class="chakra-text css-1wbddnu"
              >
                Jueves, Marzo 24, 2022
              </p>
            </div>
            <div
              class="css-k008qs"
            >
              <p
                class="chakra-text css-13ih39a"
              >
                Total:
              </p>
              <p
                class="chakra-text css-1wbddnu"
              >
                $20.00
              </p>
            </div>
            <div
              class="css-k008qs"
            >
              <p
                class="chakra-text css-13ih39a"
              >
                Código de entrega:
              </p>
              <p
                class="chakra-text css-1wbddnu"
              >
                123123
              </p>
            </div>
          </div>
          <hr
            aria-orientation="horizontal"
            class="chakra-divider css-wz74kr"
          />
          <div
            class="css-1hw8qak"
          >
            <div
              class="css-mz9qt7"
            >
              <div
                class="css-131wcsq"
              >
                <p
                  class="chakra-text css-d8f8eb"
                  data-testid="order-header"
                >
                  Estado del pedido
                </p>
              </div>
            </div>
          </div>
          <hr
            aria-orientation="horizontal"
            class="chakra-divider css-wz74kr"
          />
          <div
            class="css-f1o79g"
          >
            <div
              class="css-13ih39a"
            >
              Tus productos:
            </div>
            <div
              class="css-15qbb6"
            >
              <div
                class="css-18vhmwz"
              >
                <img
                  class="chakra-image__placeholder css-1jtd0sc"
                  src="/assets/images/noimg.png"
                />
              </div>
            </div>
            <hr
              aria-orientation="horizontal"
              class="chakra-divider css-1p98don"
            />
          </div>
          <div
            class="sc-gsDKAQ eMaAuu css-0"
          >
            <button
              class="chakra-button css-6d4n7v"
              type="button"
            >
              Ver detalle del pedido
            </button>
            <hr
              aria-orientation="vertical"
              class="chakra-divider css-12gj880"
            />
            <a
              class="chakra-link css-0"
              href="https://api.whatsapp.com/send/?phone=%2B525545439866&text=Hola%2CTengo+una+duda..&app_absent=0"
              target="_blank"
            >
              <button
                class="chakra-button css-6d4n7v"
                type="button"
              >
                Ayuda
              </button>
            </a>
          </div>
        </div>
        <hr
          aria-orientation="horizontal"
          class="chakra-divider css-18kxs9k"
        />
      </div>
    </div>
  </div>
</div>
`;
